design/design.md

```diff
@@ -0,0 +1,414 @@
+implementation shouldn't get stuck with an inefficient / broken hash function.
+Moreover, the user might want to instantiate ALBA with a PRF (e.g. AES) instead
+of a hash function modeled as a random oracle for better theoretical guarantees.
+With Rust's advanced trait system, making the hash function generic should be
+easy, and the runtime code performance should be fast due to compile time
+function resolution.
+
+Moreover, we should *require* users to provide their own hash function.
+One reason is that when ALBA is used alongside other sub-protocols, that also
+need hashing, in a larger protocol, it is the user's responsibility to implement
+[domain separation](https://en.wikipedia.org/wiki/Domain_separation)
+to provide the different sub-protocols with independent hash functions.
+Otherwise, the underlying assumptions needed for security might break and the
+larger protocol can become insecure
+(see [here](https://link.springer.com/chapter/10.1007/978-3-030-45724-2_1)).
+Additionally, different sub-protocols should not use different hash function
+constructions (e.g., SHA and BLAKE) since there might be hidden correlations
+between them and that would be equivalent to not using domain separation.
+
+Thus, the user must use a particular hash function $H$ of their choice to
+instantiate several domain separated hash functions $H_i(x) = H(i, x)$ that
+would be provided to the sub-protocols (one of them being ALBA).
+To our ALBA implementation, this domain separation would be transparent.
+
+We will still provide an example implementing a hash function with the right
+interface for ALBA using, for example, BLAKE underneath, but with a warning that
+it is provided for demonstration / prototyping only and should not be used in
+production.
+
+### Generic data elements
+Similar to the above, we should also let data elements have a generic type.
+For example, the ALBA implementation should work with a signature type that
+exposes the underlying data as a `u8` slice.
+The exact interfaces will need to be determined.
+
+### Parameter calculation
+ALBA has two kinds of parameters: user-friendly parameters such as
+$\lambda_{sec}$, $\lambda_{rel}$, $n_p$, $n_f$,
+and algorithm specific parameters such as $u$, $d$ and $q$.
+
+Our implementation will have parameter derivation code converting user-friendly
+to algorithm specific parameters.
+However, this should be decoupled from the ALBA algorithms:
+through documentation, we will ask the developer to run parameter
+derivation manually in advance and then hardcode the output algorithm specific
+parameters in their application.
+The ALBA functions such as `prove` and `verify` will thus receive as input
+the algorithm specific parameters.
+
+The rationale is that we want to be able to evolve parameter derivation
+(e.g., to optimize parameters for performance or make the derivation logic
+more rigorous) while maintaining backward compatibility with previous versions
+of the library.
+
+## Roadmap
```

Comment:
You can be more concise.

The communication with clients must disappear altogether. This is outside of the scope of this document, and is a different (non github) task to integrate Alba in Cardano's subprotocols. Likewise for the example and mathematical analysis, we should only say that the library will be maintained with research (not that we do it) and potentially enriched with more examples